arch/x86/kvm/vmx/pkvm/hyp/iommu.c

```diff
@@ -431,6 +452,9 @@ static bool sync_shadow_context_entry(struct pgt_sync_data *sdata)
 
 		pkvm_setup_ptdev_did(ptdev, context_lm_get_did(guest_ce));
 
+		if (iommu_audit_did(iommu, ptdev->did, ptdev->shadow_vm_handle))
+			return false;
+
```

Comment:
should we do did audit before set it to ptdev?

Reply:
I just used the same order of operations as in `sync_shadow_pasid_table_entry()`. But thanks for pointing it out, it seems it's indeed a good idea to do the did audit before the did setup. I'm already observing some tricky behavior related to that: when I'm starting a protected VM via crosvm with 2 devices attached, I'm observing the following sequence:

1. The 1st device is successfully attached.
2. iommu_audit_did() fails for the 2nd device, because it is not attached yet. This is an expected case explained by Chuanxiao in the comment in sync_shadow_pasid_table_entry().
3. iommu_audid_did() fails for the 1st device, because it is already attached (shadow_vm_handle != 0) while the 2nd device is still not attached (shadow_vm_handle == 0) but already has the same DID as the 1st device, since this DID was already set for the 2nd device at step 2 despite iommu_audid_did failure. This rather looks like an unexpected case.

So it looks like we should do iommu_audit_did() check before pkvm_setup_ptdev_did(), in both legacy and scalable mode, to prevent such tricky and potentially buggy situations.

Reply:
I see. With the above sequence, yes, it will break the 1st device unexpected due to it matches with 2nd device which is not attached yet. And doing iommu_audit_did() first can fix this. From my side, I didn't observe the above sequence when passthrough 2 devices to a protected VM, and the item#3 didn't happen from my side. Not sure if this is because I was using two devices belongs to two IOMMU groups.

This reminds me that, probably below sequence in host VM side may have issue. Please help to review to see if I am wrong about this. The sequence is:
1. host VM configure vIOMMU ctx/pasid entry for the passthrough device#1 with did 16(an example).
2. host VM configure vIOMMU ctx/pasid entry for the passthrough device#2 with did 16.
3. host VM trigger ctx/pasid tlb flushing, and pkvm sync_ctx/pasid for both device#1 and device#2 successfully. So they are all have did 16.
4. host VM attach device#1. pkvm sync_ctx/pasid is failed due to did audit failure (device#2 is not attahced yet but device#1 and device#2 have the same did)
5. host VM attach device#2. pkvm sync_ctx/pasid is success as both device#1 and device#2 are attached.
Sounds like the device#1 doesn't work in this sequence?

If this is true, probably this issue can be fixed by doing pkvm_iommu_sync() for each attached ptdev before populating the pgstate_pgt?

Reply:
I have to say this problem is a little tricky, host does not know the did audit, it may only want change did for pass-thru devices to same one for a guest VM. Meanwhile pKVM provide IOMMU ctx/pasid cache flush emulation, and also handle the ptdev attachment. We are doing sync_ctx/pasid and did audit during both of them. 

I am actually thinking do we must need do sync_ctx/pasid during each ptdev attaching, how about we separate it as two steps in kvm_arch_add_device_to_pkvm, first step is for hypercall to attach all ptdevs in the iommu group, them do a whole cxt/pasid cache flush?

                  dev A                  dev B
----------------------------------------------
host             did                       did            
pkvm        {did, hostVM}       {did, hostVM} <---------through ctx/pasid cache flush
                        |                            |              <--------- through ptdev attach - the issue is come from here as there is
               {did, guestVM}       {did, guestVM}               middle state for multiple ptdev support

Reply:
@cxdong :

> From my side, I didn't observe the above sequence when passthrough 2 devices to a protected VM, and the item#3 didn't happen from my side. Not sure if this is because I was using two devices belongs to two IOMMU groups.

In my case they are also in 2 different IOMMU groups.
I'm using crosvm, so if you were using qemu, the sequence might be different.
Although in this case it seems not caused by qemu vs crosvm difference. I've double-checked and found that with my patches from this PR only, I don't observe the above sequence either. But I observe it with my WIP patches for shadow page tables. The difference is that for shadow page tables I'm doing sync_shadow_context_entry() also in case of IOTLB invalidations. When I start a protected VM with 2 devices, it looks like the host does IOTLB flush for this did before attaching the 2nd device (the 1st device is already attached at this point), so the above sequence is observed.

@JasonChenCJ :

> I am actually thinking do we must need do sync_ctx/pasid during each ptdev attaching, how about we separate it as two steps in kvm_arch_add_device_to_pkvm, first step is for hypercall to attach all ptdevs in the iommu group, them do a whole cxt/pasid cache flush?

I think it's even more tricky: the devices may be in different IOMMU groups, in such case the VMM does KVM_DEV_VFIO_GROUP_ADD multiple times, so these two steps (attach all ptdevs and then whole cxt/pasid cache flush) would be not inside kvm_arch_add_device_to_pkvm() but in the VMM in userspace, i.e. we would require a certain specific behavior from userspace, which doesn't sound like a good idea.

Reply:
@semihalf-maluka-dmytro,

Yes, agree, we still need handle the case for multiple IOMMU groups, but I still would like to remove iommu sync for ptdev attachment, the thing is that 

- host want assign ptdevs to a guest, but does not expect did audit during multiple ptdev attaching, and actually such audit is not directly related with ptdev attachment.
- host will complete ptdev assignment before guest running
- before guest run, pKVM need update IOMMU ctx/pasid entry according to ptdev setting

so could an alternative way work - add one hook just before guest running, and we put all guest pre-run operations in it?
- population pgstate table
- sync IOMMU root/ctx/pasid tables
 
> I think it's even more tricky: the devices may be in different IOMMU groups, in such case the VMM does KVM_DEV_VFIO_GROUP_ADD multiple times, so these two steps (attach all ptdevs and then whole cxt/pasid cache flush) would be not inside kvm_arch_add_device_to_pkvm() but in the VMM in userspace, i.e. we would require a certain specific behavior from userspace, which doesn't sound like a good idea.

Reply:
I've updated this PR, now it does did audit before did setup (but in legacy mode only, since this PR is for legacy mode). I'll let you guys take care of fixing it for scalable mode as well as reworking it as we've discussed.